File: app/src/main/java/io/treehouses/remote/Network/BluetoothChatService.java
/*
* Copyright 2017 The Android Open Source Project, Inc.
*
* Licensed to the Apache Software Foundation (ASF) under one or more contributor
* license agreements. See the NOTICE file distributed with this work for additional
* information regarding copyright ownership. The ASF licenses this file to you under
* the Apache License, Version 2.0 (the "License"); you may not use this file except
* in compliance with the License. You may obtain a copy of the License at

* http://www.apache.org/licenses/LICENSE-2.0

* Unless required by applicable law or agreed to in writing, software distributed under
* the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
* ANY KIND, either express or implied. See the License for the specific language
* governing permissions and limitations under the License.

*/

package io.treehouses.remote.Network;

/**
 * Created by yubo on 7/11/17.
 */

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.content.Context;
import android.os.Handler;
import android.util.Log;

import com.github.ivbaranov.rxbluetooth.BluetoothConnection;
import com.github.ivbaranov.rxbluetooth.RxBluetooth;

import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Set;
import java.util.UUID;

import io.reactivex.FlowableOperator;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

import io.treehouses.remote.Constants;
import io.treehouses.remote.callback.BluetoothDeviceCallback;

/**
 * This class does all the work for setting up and managing Bluetooth
 * connections with other devices. It has a thread that listens for
 * incoming connections, a thread for connecting with a device, and a
 * thread for performing data transmissions when connected.
 */

public class BluetoothChatService implements Serializable{
    // Debugging
    private static final String TAG = "BluetoothChatService";
    // Name for the SDP record when creating server socket
    private static final String NAME_SECURE = "BluetoothChatSecure";

    // well-known SPP UUID 00001101-0000-1000-8000-00805F9B34FB
    private static final UUID MY_UUID_SECURE = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");
    private CompositeDisposable compositeDisposable;
    private BluetoothConnection bluetoothConnection;


    private static String connectedDeviceName = "NULL";
    // Member fields
    private BluetoothDevice mDevice;
    private static Handler mHandler;

    private int mCurrentState;
    private RxBluetooth bluetooth;

    private Context context;

    private Queue<String> sentCommands;
    private boolean switchedHandler = false;
    private Handler tempHandler;
    private Queue<String> tempToSendCommands;
    /**
     * Constructor. Prepares a new BluetoothChat session.
     *
     *  The UI Activity Context
     * @param handler A Handler to send messages back to the UI Activity
     */
    public BluetoothChatService(Handler handler, Context applicationContext) {
        this.context = applicationContext;
        init();
    }

    private void init() {
        compositeDisposable = new CompositeDisposable();
        mCurrentState = Constants.STATE_NONE;
        mHandler = null;
        sentCommands = new LinkedList<>();
        tempToSendCommands = new LinkedList<>();
        bluetooth = new RxBluetooth(context);
    }

    public void updateHandler(Handler handler, boolean... args){
        Log.e(TAG, "updateHandler:");
        if (!sentCommands.isEmpty() && mCurrentState == Constants.STATE_CONNECTED) {
            Log.e(TAG, "updateHandler: with overflow");
            switchedHandler = true;
            tempHandler = handler;
            tempToSendCommands.clear();
        } else {
            Log.e(TAG, "updateHandler: without overflow");
            mHandler = handler;
        }
    }

    private synchronized void updateUserInterfaceTitle() {
        Log.d(TAG, "updateUserInterfaceTitle() " + " -> " + mCurrentState);
        // Give the new state to the Handler so the UI Activity can update
        mHandler.sendMessage(mHandler.obtainMessage(Constants.MESSAGE_STATE_CHANGE, mCurrentState, -1));
    }

    public synchronized int getState() {
        return mCurrentState;
    }

    public String getConnectedDeviceName(){
        if (mDevice != null) return  mDevice.getName();
        return "NO DEVICE CONNECTED";
    }

    public void connectToDevice(BluetoothDevice device) {
        stopDiscovery();
        mCurrentState = Constants.STATE_CONNECTING;
        compositeDisposable.add(bluetooth.connectAsClient(device, MY_UUID_SECURE).subscribeOn(Schedulers.computation()).subscribe(
                bluetoothSocket -> {
                    Log.e(TAG, "connectToDevice: CONNECTED to " + device.getName());
                    mDevice = device;
                    mCurrentState = Constants.STATE_CONNECTED;
                    bluetooth.cancelDiscovery();
                    updateUserInterfaceTitle();
                    compositeDisposable.clear();
                    startChat(bluetoothSocket);
                }, throwable -> {
                    Log.e(TAG, "connectToDevice: FAILED TO CONNECT");
                    mCurrentState = Constants.STATE_FAILED;
                    updateUserInterfaceTitle();
                    init();
                }));
    }

    private void startChat(BluetoothSocket socket) throws Exception {
        Log.e(TAG, "startChat: "+"START READING" );
        bluetoothConnection = new BluetoothConnection(socket);

        compositeDisposable.add(bluetoothConnection.observeByteStream().lift((FlowableOperator<String, Byte>) this::getWriter).onBackpressureBuffer().observeOn(AndroidSchedulers.mainThread()).subscribeOn(Schedulers.io())
                .subscribe(s -> {
                    // READ COMMAND RESPONSE
                    Log.d(TAG, "READ: " + s);
                    //Remove the last command because response has been received
                    if (!sentCommands.isEmpty()) sentCommands.remove();
                    //Send the response to the target
                    if (!switchedHandler) mHandler.obtainMessage(Constants.MESSAGE_READ, s).sendToTarget();

                    //Sent all the waiting commands
                    if (switchedHandler && sentCommands.isEmpty()) {
                        Log.e(TAG, "SENT ALL PREVIOUS");
                        switchedHandler = false;
                        mHandler = tempHandler;
                        writeOverflow();
                    }

                }, throwable -> {
                    Log.e(TAG, "startChat: "+ "ERROR OCCURRED WHILE READING");
                    mCurrentState = Constants.STATE_FAILED;
                    updateUserInterfaceTitle();
                    disconnect();
                }));
    }

    private void writeOverflow() {
        while (!tempToSendCommands.isEmpty()) {
            write(tempToSendCommands.remove());
        }
    }

    public void disconnect() {
        Log.e(TAG, "DISCONNECTING");
        if (bluetoothConnection != null) {
            bluetoothConnection.closeConnection();
            bluetoothConnection = null;
            mDevice = null;
        }
        mCurrentState = Constants.STATE_NONE;
        updateUserInterfaceTitle();
        destroy();
        init();
    }

    public void startDiscovery(BluetoothDeviceCallback callback) {
        Log.e(TAG, "STARTING DISCOVERY");
        if (bluetooth.isDiscovering()) bluetooth.cancelDiscovery();
        bluetooth.startDiscovery();
        mCurrentState = Constants.STATE_LISTEN;
        updateUserInterfaceTitle();
        compositeDisposable.add(bluetooth.observeDevices()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(bluetoothDevice -> {
                    callback.onDeviceFound(bluetoothDevice);
                    Log.e(TAG, "DEVICE FOUND: "+ bluetoothDevice.getName()+ " ADDRESS: " + bluetoothDevice.getAddress());
                }));
    }

    public void stopDiscovery() {
        bluetooth.cancelDiscovery();
    }

    public Set<BluetoothDevice> getPairedDevices() {
        return bluetooth.getBondedDevices();
    }

    public boolean isBluetoothSupported() {
        return bluetooth.isBluetoothAvailable();
    }

    public boolean isBluetoothEnabled() {
        return bluetooth.isBluetoothEnabled();
    }

    private void destroy() {
        Log.e(TAG, "DESTROYING");
        if (bluetooth != null) {
            bluetooth.cancelDiscovery();
        }
        compositeDisposable.clear();
        mCurrentState = Constants.STATE_NONE;
    }

    public void write(String message) {
        if (bluetoothConnection != null) {

            Log.d(TAG, "write: " + message);
            if (switchedHandler) {
                Log.e(TAG, "WRITING TO TEMP: "+message);
                tempToSendCommands.add(message);
            } else {
                Log.e(TAG, "SENDING TO BLUETOOTH: "+message);
                sentCommands.add(message);
                bluetoothConnection.send(message);
            }
            mHandler.obtainMessage(Constants.MESSAGE_WRITE, message).sendToTarget();
        } else {
            Log.e(TAG, "Error while writing to bluetooth");
        }
    }

    private Subscriber getWriter(Subscriber subscriber) {
        return new MyWriter(subscriber);
    }

}

File: app/src/main/kotlin/io/treehouses/remote/network/BluetoothChatService.kt
package io.treehouses.remote.network

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.*
import androidx.annotation.RequiresApi
import androidx.preference.PreferenceManager
import io.treehouses.remote.Constants
import io.treehouses.remote.bases.BaseBluetoothChatService
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

class BluetoothChatService @JvmOverloads constructor(handler: Handler? = null, applicationContext: Context? = null) : BaseBluetoothChatService(handler, applicationContext) {
    inner class DisconnectReceiver: BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            val action = intent?.action
            if (action == DISCONNECT_ACTION) {
                stop()
            }
        }
    }

    private val mBinder = LocalBinder()
    private var mConnectThread: ConnectThread? = null
    private var mConnectedThread: ConnectedThread? = null
    private val receiver = DisconnectReceiver()

    @Synchronized
    override fun start() {
        bNoReconnect = false
        mConnectThread?.cancel()
        mConnectThread = null
        mConnectedThread?.cancel()
        mConnectedThread = null
        updateUserInterfaceTitle()
    }

    fun updateHandler(handler: Handler) {
        mHandler = handler
    }

    override fun onBind(intent: Intent?): IBinder {
        return mBinder
    }

    inner class LocalBinder : Binder() {
        val service: BluetoothChatService
            get() = this@BluetoothChatService
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        return START_NOT_STICKY
    }

    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
    override fun onCreate() {
        super.onCreate()
        context = applicationContext
        val i = IntentFilter()
        i.addAction(DISCONNECT_ACTION)
        registerReceiver(receiver, i, RECEIVER_NOT_EXPORTED)
    }

    override fun onDestroy() {
        stop()
        unregisterReceiver(receiver)
        super.onDestroy()
    }

    var connectedDeviceName: String? = ""

    @Synchronized
    fun connect(device: BluetoothDevice?, secure: Boolean) {
        if (state == Constants.STATE_CONNECTING) {
            if (mConnectThread != null) {
                mConnectThread?.cancel()
                mConnectThread = null
            }
        }
        if (mConnectedThread != null) {
            mConnectedThread?.cancel()
            mConnectedThread = null
        }

        mConnectThread = ConnectThread(device, secure)
        mConnectThread?.start()
        updateUserInterfaceTitle()
    }

    @Synchronized
    fun connected(socket: BluetoothSocket?, device: BluetoothDevice?) {
        connectedDeviceName = device?.name
        mDevice = device
        if (mConnectThread != null) {
            mConnectThread?.cancel()
            mConnectThread = null
        }

        if (mConnectedThread != null) {
            mConnectedThread?.cancel()
            mConnectedThread = null
        }

        startNotification()
        mConnectedThread = ConnectedThread(socket)
        mConnectedThread?.start()

        updateUserInterfaceTitle()
        val msg = mHandler?.obtainMessage(Constants.MESSAGE_DEVICE_NAME)
        val bundle = Bundle()
        bundle.putString(Constants.DEVICE_NAME, device?.name)
        msg?.data = bundle
        mHandler?.sendMessage(msg ?: Message())
    }

    @Synchronized
    fun stop() {
        bNoReconnect = true
        if (mConnectThread != null) {
            mConnectThread?.cancel()
            mConnectThread = null
        }
        if (mConnectedThread != null) {
            mConnectedThread?.cancel()
            mConnectedThread = null
        }

        state = Constants.STATE_NONE
        stopForeground(true)
        updateUserInterfaceTitle()
    }

    fun write(out: ByteArray?) {
        var r: ConnectedThread?
        synchronized(this) {
            if (state != Constants.STATE_CONNECTED) return
            r = mConnectedThread
        }
        if (out != null) {
            r?.write(out)
        }
    }

    private fun connectionLost() {
        callHandler("Device connection was lost")
        stopForeground(true)

        val preferences = PreferenceManager.getDefaultSharedPreferences(applicationContext)
        if (mDevice != null && !bNoReconnect && preferences.getBoolean("reconnectBluetooth", true)) {
            connect(mDevice, true)
        } else {
            state = Constants.STATE_NONE
            updateUserInterfaceTitle()
            start()
        }
    }

    private inner class ConnectThread(private val mmDevice: BluetoothDevice?, secure: Boolean) : Thread() {
        private val mmSocket: BluetoothSocket?
        private val mSocketType: String
        override fun run() {
            name = "ConnectThread$mSocketType"
            this@BluetoothChatService.state = Constants.STATE_CONNECTING
            mAdapter?.cancelDiscovery()
            try {
                mmSocket?.connect()
            } catch (e: Exception) {
                e.printStackTrace()
                closeSocket()
                connectionFailed()
                return
            }
            synchronized(this@BluetoothChatService) { mConnectThread = null }
            connected(mmSocket, mmDevice)
        }

        fun cancel() { closeSocket() }

        fun closeSocket() {
            try { mmSocket?.close() }
            catch (e: Exception) { e.printStackTrace() }
        }

        init {
            var tmp: BluetoothSocket? = null
            mSocketType = if (secure) "Secure" else "Insecure"
            try {
                tmp = mmDevice?.createRfcommSocketToServiceRecord(MY_UUID_SECURE)
                this@BluetoothChatService.state = Constants.STATE_CONNECTING
            } catch (e: Exception) {
                e.printStackTrace()
                this@BluetoothChatService.state = Constants.STATE_NONE
            }
            this@BluetoothChatService.updateUserInterfaceTitle()
            mmSocket = tmp
        }
    }

    private inner class ConnectedThread(socket: BluetoothSocket?) : Thread() {
        private val mmSocket: BluetoothSocket? = socket
        private val mmInStream: InputStream?
        private val mmOutStream: OutputStream?
        override fun run() {
            val buffer = ByteArray(10000)
            var bytes: Int
            var out: String
            while (true) {
                try {
                    bytes = mmInStream?.read(buffer) ?: 0
                    out = String(buffer, 0, bytes)
                    mHandler?.obtainMessage(Constants.MESSAGE_READ, bytes, -1, out)?.sendToTarget()
                } catch (e: IOException) {
                    e.printStackTrace()
                    connectionLost()
                    break
                }
            }
        }

        fun write(buffer: ByteArray) {
            try {
                mmOutStream?.write(buffer)
                mHandler?.obtainMessage(Constants.MESSAGE_WRITE, -1, -1, buffer)?.sendToTarget()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

        fun cancel() {
            try {
                mmSocket?.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        init {
            var tmpIn: InputStream? = null
            var tmpOut: OutputStream? = null
            try {
                tmpIn = socket?.inputStream
                tmpOut = socket?.outputStream
            } catch (e: IOException) {
                e.printStackTrace()
            }
            mmInStream = tmpIn
            mmOutStream = tmpOut
            this@BluetoothChatService.state = Constants.STATE_CONNECTED
            this@BluetoothChatService.updateUserInterfaceTitle()
        }
    }
}